Refactor: Modular operator polling, hot reload support, structured logging

ciospettw opened the pull request:
tutti i cambiamenti sono motivati, se manca qualcosa possiamo discuterne domani in call.

Provato anche in locale tramite server redis e funziona tutto

ciospettw committed:
Update LICENSE

aggiunta crediti

ciospettw committed:
Update config.example.json

ciospettw committed:
Update index.js

ciospettw committed:
Update package-lock.json

ciospettw committed:
Update package.json

ciospettw committed:
Update redisClient.js

ciospettw committed:
Update tokenManager.js

ciospettw committed:
Update tools.js

ciospettw committed:
Update utils.js

ciospettw committed:
Update README.md

ciospettw committed:
Update README.md

ciospettw committed:
Update README.md

Giasinit commented:
Ciao @ciospettw,
ho guardato un po’ il codice e devo dire che mi piace molto l’approccio che hai preso.. la modularizzazione con i plugin per ogni operatore è super utile.. anche il logging strutturato e il reload della config senza riavvio sono davvero comodi, soprattutto in produzione.. bravo!

Ci tenevo solo a segnalare che nel caso di BIV (Busitalia Veneto) i feed bus & tram sono separati.. rif. https://gtfs-biv.fsbusitalia.com/ , mi chiedevo se avevi già previsto questa cosa. Grazie

Giasinit commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2025 Giasin
+Copyright (c) 2025 Giasin / fede
```
invece di `Copyright (c) 2025 Giasin / fede` possiamo mettere `Copyright (c) 2025 PadoVia and its contributors`.. che ne dici?

ciospettw replied:
> invece di `Copyright (c) 2025 Giasin / fede` possiamo mettere `Copyright (c) 2025 PadoVia and its contributors`.. che ne dici?

https://github.com/PadoVia/tpl-avl-daemon/pull/1#discussion_r2173548393

Giasinit commented on config.example.json:
```diff
@@ -30,17 +30,11 @@
-          "method": "GET"
-        },
-        "bus": {
-          "url": "https://gtfs.example.org/vehicle-positions-bus.pb",
```
la cosa che ti dicevo prima: https://github.com/PadoVia/tpl-avl-daemon/pull/1#issuecomment-3016098790

Giasinit commented on package.json:
```diff
@@ -1,38 +1,26 @@
+      "start": "node index.js",
+      "dev": "nodemon index.js"
+    },
+    "author": "fede",
```
mettiamo `PadoVia and its contributors`?

ciospettw replied:
per ora va bene, ma se va in deployment i crediti devono essere necessariamente piu specifici poiche non faccio parte vera e propria nel progetto - un tag separato mi starebbe piu simpatico haha ^-^

ciospettw replied:
> mettiamo `PadoVia and its contributors`?

per ora va bene, ma se va in deployment i crediti devono essere necessariamente piu specifici poiche non faccio parte vera e propria nel progetto - un tag separato mi starebbe piu simpatico haha ^-^

Giasinit replied:
va bene va bene

Giasinit commented on tokenManager.js:
```diff
@@ -1,52 +1,73 @@
-  tokens.delete(slug);
-}
+    try {
+        const res = await axios({
```
Ovviamente questo è fatto per BIV.. però in futuro potrebbe cambiare ;)

Giasinit commented on utils.js:
```diff
@@ -1,73 +1,99 @@
   throw lastErr;
 }
 
+// converte timestamp Roma (locale) in ISO UTC (vedo che era particolarmente richiesto)
```
Si.. il AVL di BIV fornisce il timestamp in Europe/Rome

Giasinit commented on utils.js:
```diff
@@ -1,73 +1,99 @@
+  return (θ + 360) % 360;
 }
 
+// calcola velocità media tra due pos/time point (in km/h) - ok, non é male.. ma di solito l'rt include la speed. usate quella, direi, per poi 
```
peccato che il GTFS-RT di BIV non abbia questa cosa.. poi come dicevo.. l'AVL ha lo speed.. ma se usiamo anche i dati del GTFS-RT è giusto poter calcolare qualcosa(giusto come fallback).

ciospettw replied:
si certo, la funzione la ho lasciata la stessa per un motivo (sistemata perche ogni step era sotto l'altro(non c'e bisogno), se vuoi falla ritornare come prima, anche se cosi lo vedo piu ordinato) :)

ciospettw commented:
> Ciao @ciospettw, ho guardato un po’ il codice e devo dire che mi piace molto l’approccio che hai preso.. la modularizzazione con i plugin per ogni operatore è super utile.. anche il logging strutturato e il reload della config senza riavvio sono davvero comodi, soprattutto in produzione.. bravo!
> 
> Ci tenevo solo a segnalare che nel caso di BIV (Busitalia Veneto) i feed bus & tram sono separati.. rif. https://gtfs-biv.fsbusitalia.com/ , mi chiedevo se avevi già previsto questa cosa. Grazie

l’architettura dell’app (come anche lo era prima delle modifiche) è pensata per essere molto modulare: ogni operatore dispone di un proprio modulo nella cartella /operators e può essere configurato tramite una voce dedicata nel config 

per casi come BIV dove i feed bus e tram sono separati, beh, l’approccio attuale sarebbe di definire due operatori distinti nella configurazione: uno per il feed bus e uno per il feed tram, ciascuno con il proprio endpoint GTFS-RT/avl

se si volesse invece supportare più feed per singolo operatore, bisognerebbe sia modificare la struttura della configurazione (trasformare gtfsrt/avl da oggetto a array di feed distinti, ciascuno con le sue credenziali e URL) e modificare la logica di polling nel codice per ciclare questi feed e gestirli in maniera isolata (sia nel polling che nei log).

lo posso fare.. chiaramente non ora lol, ma se vuoi oggi giornata mi ci metto.

Giasinit commented:
> > Ciao @ciospettw, ho guardato un po’ il codice e devo dire che mi piace molto l’approccio che hai preso.. la modularizzazione con i plugin per ogni operatore è super utile.. anche il logging strutturato e il reload della config senza riavvio sono davvero comodi, soprattutto in produzione.. bravo!
> > Ci tenevo solo a segnalare che nel caso di BIV (Busitalia Veneto) i feed bus & tram sono separati.. rif. https://gtfs-biv.fsbusitalia.com/ , mi chiedevo se avevi già previsto questa cosa. Grazie
> 
> l’architettura dell’app (come anche lo era prima delle modifiche) è pensata per essere molto modulare: ogni operatore dispone di un proprio modulo nella cartella /operators e può essere configurato tramite una voce dedicata nel config
> 
> per casi come BIV dove i feed bus e tram sono separati, beh, l’approccio attuale sarebbe di definire due operatori distinti nella configurazione: uno per il feed bus e uno per il feed tram, ciascuno con il proprio endpoint GTFS-RT/avl
> 
> se si volesse invece supportare più feed per singolo operatore, bisognerebbe sia modificare la struttura della configurazione (trasformare gtfsrt/avl da oggetto a array di feed distinti, ciascuno con le sue credenziali e URL) e modificare la logica di polling nel codice per ciclare questi feed e gestirli in maniera isolata (sia nel polling che nei log).
> 
> lo posso fare.. chiaramente non ora lol, ma se vuoi oggi giornata mi ci metto.

Grazie della spiegazione, quando vuoi e quando riesci fai pure.. non è urgente per ora.. sopratutto perché da noi il tram viene sospeso dal 30 giugno 2025 al 10 settembre 2025..

ciospettw committed:
Update config.example.json

supporti multipli in file config

ciospettw committed:
Update index.js

coerenza per support multi feed x op

ciospettw committed:
Update tools.js

coerenza per supporto tools.js multi feed x op

ciospettw committed:
Update busitalia_veneto_padua.js

refactor: support multi-feed; mantiene handler GTFSRT/SIRI per coerenza

ciospettw committed:
Create example_operator.js

docs: aggiunto esempio per operatori multi-feed AVL/GTFSRT/SIRI

Giasinit commented on operators/busitalia_veneto_padua.js:
```diff
@@ -81,73 +84,44 @@ async function fetchVehicles(token, config) {
+    const v = entity.vehicle;
+    return {
+      plate: v.vehicle ? v.vehicle.label : null,
+      timestamp: v.timestamp ? v.timestamp.low : null,
```
il nostro standard è usare ISOString.. per evitare il unix timestamp..

Giasinit commented:
Comunque, ora provo con i dati di BIV

ciospettw committed:
Update index.js

refactor n3: ripristino di tutte le funzioni di logging avanzato, hot reload, express e polling multi-feed che avevo rimosso per far vedere la logica in modo chiaro

ciospettw committed:
Update tools.js

refactor n3: aggiunto supporto multi-feed alla funzione fetchVehiclesForOperator, riaggiunta di tutte le funzioni depletate

ciospettw committed:
Update config.example.json

patch per chiarezza del config

ciospettw committed:
Update busitalia_veneto_padua.js

refactor: reintrodotta funzione avanzata GTFS-RT rimossa temporaneamente perché non sapevo se sarebbe crashata col multi-feed; ho testato e sembra essere tutto compatibile.

ciospettw commented:
> Comunque, ora provo con i dati di BIV

certo. comunque ho fatto nuove modifiche perche avevo tolto delle funzioni, non sapevo se sarebbero crashate col multi feed. 
sembrerebbe funzionare tutto.

Giasinit committed:
fix: add missing await and rename variable for clarity

Giasinit committed:
fix(token): use first AVL feed in array as fallback for unhandled case

Giasinit committed:
feat: pass full feed object to fetchVehiclesGTFSRT

Updated the GTFS-RT vehicle fetch logic to pass the entire feed object instead of just the URL. This allows fetchVehiclesGTFSRT to access additional feed properties if needed.

Giasinit commented:
ho controllato e fatto i miei test.. mi sembra tutto apposto per ora.. quindi posso mergiare.
